File: javascript/robot-name/robot-name.js
function Robot() {
    this.reset();
};

Robot.prototype.reset = function() {
    while (true) {
        name = this.generateName();
        if (!this._names[name]) {
            this._names[name] = true;
            this.name = name;
            break;
        }
    }
}

Robot.prototype._names = {};

Robot.prototype.generateName = function() {
    return this.randomChar() + this.randomChar() + this.randomIntStr();
}

Robot.prototype.randomChar = function() {
    return String.fromCharCode(parseInt(Math.random() * 100) % 26 + 65);
};

Robot.prototype.randomIntStr = function() {
    s = '000' + parseInt(Math.random() * 1000).toString();
    return s.substr(s.length - 3);;
}

module.exports = Robot;

File: cpp/robot-name/robot_name.cpp
#include "robot_name.h"
#include <sstream>


namespace robot_name
{

using namespace std;


robot::robot() 
{
    reset();
}

const string& robot::name() const
{
    return _name;
}

void robot::reset()
{
    do
    {
        ostringstream os;
        os << _alpha() << _alpha() << _digit() << _digit() << _digit();
        _name = os.str();
    } while (_names.count(_name));
    
    _names.insert(_name);
}

random_device robot::_rd{};
set<string> robot::_names{};

const char robot::_alpha()
{
    return _rd() % 26 + 'A';
}

const char robot::_digit()
{
    return _rd() % 10 + '0';
}

} // namespace robot_name